phoenix-core/src/main/java/org/apache/phoenix/coprocessor/IndexerRegionScanner.java

```diff
@@ -305,42 +320,96 @@ private void verifyIndex() throws IOException {
         Cell lastCell = null;
         int rowCount = 0;
         region.startRegionOperation();
         try {
             synchronized (innerScanner) {
+                byte[] uuidValue = ServerCacheClient.generateId();
                 do {
                     List<Cell> row = new ArrayList<>();
                     hasMore = innerScanner.nextRaw(row);
                     if (!row.isEmpty()) {
-                        lastCell = row.get(0);
+                        lastCell = row.get(0); // lastCell is any cell from the last visited row
                         Put put = null;
+                        Delete del = null;
                         for (Cell cell : row) {
                             if (KeyValue.Type.codeToType(cell.getTypeByte()) == KeyValue.Type.Put) {
+                                if (!partialRebuild && familyMap != null && !isColumnIncluded(cell)) {
+                                    continue;
+                                }
                                 if (put == null) {
                                     put = new Put(CellUtil.cloneRow(cell));
                                 }
                                 put.add(cell);
                             } else {
-                                throw new DoNotRetryIOException("Scan without raw found a deleted cell");
+                                if (del == null) {
+                                    del = new Delete(CellUtil.cloneRow(cell));
+                                }
+                                del.addDeleteMarker(cell);
+                            }
+                        }
+                        if (put == null && del == null) {
+                            continue;
+                        }
+                        if (!verify) {
+                            if (put != null) {
+                                setMutationAttributes(put, uuidValue);
+                                mutations.add(put);
                             }
+                            if (del != null) {
+                                setMutationAttributes(del, uuidValue);
+                                mutations.add(del);
+                            }
+                            uuidValue = commitIfReady(uuidValue, mutations);
+                        } else {
+                            indexKeyToDataPutMap
+                                    .put(getIndexRowKey(indexMaintainer, put), put);
                         }
                         rowCount++;
-                        indexKeyToDataPutMap
-                                .put(getIndexRowKey(indexMaintainer, put), put);
+
                     }
                 } while (hasMore && rowCount < pageSizeInRows);
-                verifyIndex();
+                if (verify) {
+                    verifyIndex();
+                } else if (!mutations.isEmpty()) {
+                    ungroupedAggregateRegionObserver.checkForRegionClosing();
```

Comment:
I'm a bit confused -- I thought IndexerRegionScanner was only used for old-style indexes when doing ONLY-mode index verifications so we got the new correct verify features on the old indexes -- why are we committing things?

Reply:
OK, I see below that we're now using IndexerRegionScanner more broadly. Is this necessary -- just curious if it's worth doing a lot of refactoring and requisite testing on an index framework we know we want to deprecate soon. If it's actually harder to avoid the refactoring while touching UARO and the new indexing classes, that's fine.

Reply:
This PR moved the rebuild path from IndexRegionObserver into UARO, more specifically IndexRebuildRegionObserver. This means that we cannot use IndexRebuildRegionObserver for building old indexes anymore. So I needed to move the responsibility of building old indexes to IndexerRegionObserver in order to continue using the index rebuild code in Indexer for old indexes.